packages/app/src/lib/1pux-parser.ts

```diff
@@ -0,0 +1,308 @@
+};
+
+export const parse1PuxFile = async (
+  fileContents: string | ArrayBuffer,
+) => {
+  try {
+    const zip = await loadAsync(fileContents);
+
+    const attributesContent = await zip
+      .file('export.attributes')!
+      .async('string');
+    const attributes = JSON.parse(attributesContent);
+    const dataContent = await zip.file('export.data')!.async('string');
+    const data = JSON.parse(dataContent);
+
+    return {
+      attributes,
+      data,
+    } as OnePuxExport;
+  } catch (error) {
+    console.error('Failed to parse .1pux file');
+    throw error;
+  }
+};
+
+type RowData = {
+  name: string;
+  tags: string;
+  url: string;
+  username: string;
+  password: string;
+  notes: string;
+  extraFields: ExtraField[];
+};
+
+type ExtraFieldType =
+  | 'username'
+  | 'password'
+  | 'url'
+  | 'email'
+  | 'date'
+  | 'month'
+  | 'credit'
+  | 'phone'
+  | 'totp'
+  | 'text';
+
+type ExtraField = { name: string; value: string; type: ExtraFieldType };
+
+type ParseFieldTypeToExtraFieldType = (
+  field: OnePuxItemDetailsLoginField,
+) => ExtraFieldType;
+
+const parseFieldTypeToExtraFieldType: ParseFieldTypeToExtraFieldType = (
+  field,
+) => {
+  if (field.designation === 'username') {
+    return 'username';
+  } else if (field.designation === 'password') {
+    return 'password';
+  } else if (field.fieldType === 'E') {
+    return 'email';
+  } else if (field.fieldType === 'U') {
+    return 'url';
+  }
+  return 'text';
+};
+
+export const parseToRowData = (
+  item: OnePuxItem['item'],
+  defaultTags?: string[],
+) => {
+  if (!item) {
+    return;
+  }
+
+  const rowData: RowData = {
+    name: item.overview.title,
+    tags: [...(defaultTags || []), ...(item.overview.tags || [])].join(','),
+    url: item.overview.url || '',
+    username: '',
+    password: '',
+    notes: item.details.notesPlain || '',
+    extraFields: [],
+  };
+
+  // Skip documents
+  if (
+    item.details.documentAttributes &&
+    item.details.loginFields.length === 0
+  ) {
+    return;
+  }
+
+  // Extract username, password, and some extraFields
+  item.details.loginFields.forEach((field) => {
+    if (field.designation === 'username') {
+      rowData.username = field.value;
+    } else if (field.designation === 'password') {
+      rowData.password = field.value;
+    } else if (
+      field.fieldType === 'I' ||
+      field.fieldType === 'C' ||
+      field.id.includes(';opid=__') ||
+      field.value === ''
+    ) {
+      // Skip these noisy form-fields
+      return;
+    } else {
+      rowData.extraFields.push({
+        name: field.name || field.id,
+        value: field.value,
+        type: parseFieldTypeToExtraFieldType(field),
+      });
+    }
+  });
+
+  // Extract some more extraFields
+  item.details.sections.forEach((section) => {
+    section.fields.forEach((field) => {
+      let value = '';
+      let type: ExtraFieldType = 'text';
+
+      if (Object.prototype.hasOwnProperty.call(field.value, 'concealed')) {
```

Comment:
I generally go with this safer-if-more-verbose approach on frontend code (more easily tampered with), but we don't have to. You can see some info on why at https://eslint.org/docs/rules/no-prototype-builtins

Reply:
In this case (and probably in most other cases as well) I would prefer the more terse, readable version unless there is a concrete failure mode we're trying to avoid. In this case I don't see any real risk in simply checking for thruthyness.